src/parser/core/Entity.ts

```diff
@@ -211,27 +211,31 @@ class Entity {
     sourceId: number | null = null,
   ): Map<number, number> {
     const buffTimesAtApplication: Map<number, number> = new Map<number, number>();
-    const sortedApplicationTimestamps: number[] = this.getBuffHistory(spellId, sourceId)
-      .map((buffEvent) => buffEvent.timestamp)
-      .sort((a, b) => a - b);
+
+    const sortedApplicationTimestamps: number[] = this.getBuffHistory(spellId, sourceId).flatMap(
+      (e) => [e.timestamp, ...e.refreshHistory],
+    );
+
+    sortedApplicationTimestamps.sort((a, b) => a - b);
+
     if (sortedApplicationTimestamps.length === 0) {
       return buffTimesAtApplication;
     }
     let lastApplicationTimestamp: number = sortedApplicationTimestamps[0];
     let buffTimeAtLastApplication: number = baseBuffLength;
-    buffTimesAtApplication.set(lastApplicationTimestamp, 0);
+    buffTimesAtApplication.set(lastApplicationTimestamp, buffTimeAtLastApplication);
     for (let i = 1; i < sortedApplicationTimestamps.length; i += 1) {
       const timeDiffBetweenApplications = sortedApplicationTimestamps[i] - lastApplicationTimestamp;
       const buffAmountAtCurrentApplication = Math.max(
         0,
         buffTimeAtLastApplication - timeDiffBetweenApplications,
       );
-      buffTimesAtApplication.set(sortedApplicationTimestamps[i], buffAmountAtCurrentApplication);
       lastApplicationTimestamp = sortedApplicationTimestamps[i];
       buffTimeAtLastApplication = Math.min(
         buffAmountAtCurrentApplication + baseBuffLength,
         maxBuffLength,
       );
+      buffTimesAtApplication.set(sortedApplicationTimestamps[i], buffTimeAtLastApplication);
```

Comment:
my understanding of the changes to this file is:

- you added refreshes to `sortedApplicationTimestamps`, so responses from this method are now conceptually "times a buff would have been applied" and not just "times a buff was actually freshly applied"
- the change to the first apply to give it the full `baseBuffLength` remaining duration seems fine
- the change here handles pandemic-ing a buff.

is that correct?

Reply:
Yes. That is correct. From what I can tell, the one usage of this function (getRemainingBuffTimeAtTimestamp below) seems to be implemented with this logic in mind.